# Importing libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing data

In [2]:
# Turn your project folder path into a string
path = r'/Users/dianafayarbas/Desktop/10.24.2023 Instacart Basket Analysis'

In [3]:
path

'/Users/dianafayarbas/Desktop/10.24.2023 Instacart Basket Analysis'

In [4]:
# Import your main orders_products_merged dataframe 
# (the one you exported after completing the task in the previous Exercise)
# Call it ords_prods_merge in your new notebook

ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_4.7update.pkl'))

In [5]:
# Create a subset called df of the first one million entries
df = ords_prods_merge[:1000000]
df.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered     busiest_day  \
0                     3.0                  5          0  Regularly busy   
1                    20.0                  1          1  Regularly busy   
2                     6.0                 20          0     Busiest day   
3                     NaN                 10          0  Regularly busy   
4                     9.0                 11          1      Least busy   

     price_range_loc    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy           Most orders  
1  Mid-range product  Regularly busy        Average orders  
2  Mid-range product    Busiest days        Average orders  
3  Mid-range product      Least busy           Most orders  
4  Mid-range product      Least busy        Average orders

In [6]:
df.shape

(1000000, 17)

# Grouping Data with pandas

In [7]:
df.groupby('product_name')

# Aggregating Data with agg()

In [8]:
# Average number of orders per user for each department ID
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

In [9]:
# Achieve the same results
df.groupby('department_id')['order_number'].mean()

department_id
1     15.577493
2     17.320781
3     16.084944
4     17.530458
5     14.763075
6     16.658449
7     17.031590
8     15.076662
9     15.447580
10    18.681852
11    15.447411
12    14.327957
13    16.548642
14    16.960241
15    16.121948
16    17.803851
17    15.593633
18    19.674252
19    16.899756
20    16.255442
21    25.535479
Name: order_number, dtype: float64

In [10]:
# Multiple statistics
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.577493   1  99
2                17.320781   1  96
3                16.084944   1  99
4                17.530458   1  99
5                14.763075   1  99
6                16.658449   1  99
7                17.031590   1  99
8                15.076662   1  98
9                15.447580   1  99
10               18.681852   1  99
11               15.447411   1  99
12               14.327957   1  99
13               16.548642   1  99
14               16.960241   1  99
15               16.121948   1  99
16               17.803851   1  99
17               15.593633   1  99
18               19.674252   1  99
19               16.899756   1  99
20               16.255442   1  99
21               25.535479   1  98

# Aggregating Data with transform()

In [11]:
# Create your “max_order” column
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)

In [12]:
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered     busiest_day  \
0                     3.0                  5          0  Regularly busy   
1                    20.0                  1          1  Regularly busy   
2                     6.0                 20          0     Busiest day   
3                     NaN                 10          0  Regularly busy   
4                     9.0                 11          1      Least busy   

     price_range_loc    busiest_days busiest_period_of_day  max_order  
0  Mid-range product  Regularly busy           Most orders         32  
1  Mid-range product  Regularly busy        Average orders         32  
2  Mid-range product    Busiest days        Average orders          5  
3  Mid-range product      Least busy           Most orders          3  
4  Mid-range product      Least busy        Average orders          3

In [13]:
# pd.options.display.max_rows = None
# ords_prods_merge.head(100)

# Deriving Columns with loc()

In [14]:
# Create a flag that assigns a “loyalty” label to a user ID based on 
# its corresponding max order value

ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
# Print the frequency of your new “loyalty_flag” column
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [18]:
# Write a head() function that returned the first 60 rows of only the “user_id,” 
# “loyalty_flag,” and “order_number” columns
# ords_prods_merge[['user_id', 'loyalty_flag', 'order_number']].head(60)

# Task 4.8

### Step 1

In [19]:
# Create a new notebook for this task. Be sure to import the relevant libraries, 
# along with your ords_prods_merge dataframe, which should include your newly derived 
# columns from the previous Exercise.
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered     busiest_day  \
0                     3.0                  5          0  Regularly busy   
1                    20.0                  1          1  Regularly busy   
2                     6.0                 20          0     Busiest day   
3                     NaN                 10          0  Regularly busy   
4                     9.0                 11          1      Least busy   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         32   
1  Mid-range product  Regularly busy        Average orders         32   
2  Mid-range product    Busiest days        Average orders          5   
3  Mid-range product      Least busy           Most orders          3   
4  Mid-range product      Least busy        Average orders          3   

       loyalty_flag  
0  Regular customer  
1  Regular customer  
2      New customer  
3      New customer  
4      New customer

### Step 2

In [20]:
# In this Exercise, you learned how to find the aggregated mean of the “order_number” column
# grouped by “department_id” for a subset of your dataframe. Now, repeat this process for 
# the entire dataframe.

# Average number of orders per user for each department ID
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

In [21]:
# Running df again for comparison
df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.577493
2                17.320781
3                16.084944
4                17.530458
5                14.763075
6                16.658449
7                17.031590
8                15.076662
9                15.447580
10               18.681852
11               15.447411
12               14.327957
13               16.548642
14               16.960241
15               16.121948
16               17.803851
17               15.593633
18               19.674252
19               16.899756
20               16.255442
21               25.535479

### Step 3

In [22]:
# Analyze the result. How do the results for the entire dataframe differ from those of
# the subset? Include your comments in a markdown cell below the executed code.

**The order_number means for the whole dataset and for the subset are different. For example, the whole dataset's order_number mean for department 3 is larger (17.170) than the subset's order_number mean for the same department (16.085).**

### Step 4

In [23]:
# Follow the instructions in the Exercise for creating a loyalty flag for existing customers 
# using the transform() and loc() functions.

# I already used transform() to create a “max_order” column
# ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.max)
# ords_prods_merge.head()

In [24]:
# I already created a flag that assigns a “loyalty” label to a user ID based on 
# its corresponding max order value
# Here is the code again:

# ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [25]:
# ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [26]:
# ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [27]:
# Dataframe already contains max_order and loyalty_flag columns
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered     busiest_day  \
0                     3.0                  5          0  Regularly busy   
1                    20.0                  1          1  Regularly busy   
2                     6.0                 20          0     Busiest day   
3                     NaN                 10          0  Regularly busy   
4                     9.0                 11          1      Least busy   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         32   
1  Mid-range product  Regularly busy        Average orders         32   
2  Mid-range product    Busiest days        Average orders          5   
3  Mid-range product      Least busy           Most orders          3   
4  Mid-range product      Least busy        Average orders          3   

       loyalty_flag  
0  Regular customer  
1  Regular customer  
2      New customer  
3      New customer  
4      New customer

### Step 5

In [28]:
# Use the loyalty flag you created and check the basic statistics of the product prices 
# for each loyalty category (Loyal Customer, Regular Customer, and New Customer). 
# What you’re trying to determine is whether the prices of products purchased by 
# loyal customers differ from those purchased by regular or new customers.

ords_prods_merge.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

### Step 6

In [29]:
# Create a spending flag for each user based on the average price across all their orders
# If the mean of the prices of products purchased by a user is lower than 10, 
# then flag them as a “Low spender.”
# If the mean of the prices of products purchased by a user is higher than or equal to 10, 
# then flag them as a “High spender.”

# Create an "avg_order” column
ords_prods_merge['avg_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform(np.mean)
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_prior_order  add_to_cart_order  reordered     busiest_day  \
0                     3.0                  5          0  Regularly busy   
1                    20.0                  1          1  Regularly busy   
2                     6.0                 20          0     Busiest day   
3                     NaN                 10          0  Regularly busy   
4                     9.0                 11          1      Least busy   

     price_range_loc    busiest_days busiest_period_of_day  max_order  \
0  Mid-range product  Regularly busy           Most orders         32   
1  Mid-range product  Regularly busy        Average orders         32   
2  Mid-range product    Busiest days        Average orders          5   
3  Mid-range product      Least busy           Most orders          3   
4  Mid-range product      Least busy        Average orders          3   

       loyalty_flag  avg_order  
0  Regular customer  14.790541  
1  Regular customer  14.790541  
2      New customer   3.062500  
3      New customer   2.017241  
4      New customer   2.017241

In [30]:
# Create a spending flag
ords_prods_merge.loc[ords_prods_merge['avg_order'] < 10, 'spending_flag'] = 'Low spender'

In [31]:
ords_prods_merge.loc[ords_prods_merge['avg_order'] >= 10, 'spending_flag'] = 'High spender'
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  add_to_cart_order  reordered     busiest_day    price_range_loc  \
0  ...                  5          0  Regularly busy  Mid-range product   
1  ...                  1          1  Regularly busy  Mid-range product   
2  ...                 20          0     Busiest day  Mid-range product   
3  ...                 10          0  Regularly busy  Mid-range product   
4  ...                 11          1      Least busy  Mid-range product   

     busiest_days busiest_period_of_day max_order      loyalty_flag  \
0  Regularly busy           Most orders        32  Regular customer   
1  Regularly busy        Average orders        32  Regular customer   
2    Busiest days        Average orders         5      New customer   
3      Least busy           Most orders         3      New customer   
4      Least busy        Average orders         3      New customer   

   avg_order  spending_flag  
0  14.790541   High spender  
1  14.790541   High spender  
2   3.062500    Low spender  
3   2.017241    Low spender  
4   2.017241    Low spender  

[5 rows x 21 columns]

### Step 7

In [32]:
# Create an order frequency flag that marks the regularity of a user’s ordering behavior 
# according to the median in the “days_since_prior_order” column. 

# Create an "median_dspo” column
ords_prods_merge['median_dspo'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.median)
ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  reordered     busiest_day    price_range_loc    busiest_days  \
0  ...          0  Regularly busy  Mid-range product  Regularly busy   
1  ...          1  Regularly busy  Mid-range product  Regularly busy   
2  ...          0     Busiest day  Mid-range product    Busiest days   
3  ...          0  Regularly busy  Mid-range product      Least busy   
4  ...          1      Least busy  Mid-range product      Least busy   

  busiest_period_of_day max_order      loyalty_flag  avg_order spending_flag  \
0           Most orders        32  Regular customer  14.790541  High spender   
1        Average orders        32  Regular customer  14.790541  High spender   
2        Average orders         5      New customer   3.062500   Low spender   
3           Most orders         3      New customer   2.017241   Low spender   
4        Average orders         3      New customer   2.017241   Low spender   

   median_dspo  
0          8.0  
1          8.0  
2          8.0  
3          9.0  
4          9.0  

[5 rows x 22 columns]

In [33]:
# Create an order frequency flag
# If the median of “days_since_prior_order” is higher than 20, 
# then the customer should be labeled a “Non-frequent customer.”
ords_prods_merge.loc[ords_prods_merge['median_dspo'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [34]:
# If the median is higher than 10 and lower than or equal to 20, 
# then the customer should be labeled a “Regular customer.”
ords_prods_merge.loc[(ords_prods_merge['median_dspo'] > 10) & (ords_prods_merge['median_dspo'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [35]:
# If the median is lower than or equal to 10, 
# then the customer should be labeled a “Frequent customer.”
ords_prods_merge.loc[ords_prods_merge['median_dspo'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [36]:
ords_prods_merge['order_freq_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_freq_flag, dtype: int64

### Step 9

In [38]:
# Export your dataframe as a pickle file and store it correctly
# in your “Prepared Data” folder.

ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_4.8update.pkl'))